compiler/optimizations/inferConstRefs.cpp

```diff
@@ -103,163 +150,62 @@ static bool canRHSBeConstRef(CallExpr* parent, CallExpr* rhs, SymExpr* use) {
-    case PRIM_TYPE_INIT:
-    case PRIM_LOGICAL_FOLDER:
-    case PRIM_TUPLE_EXPAND:
-    case PRIM_TUPLE_AND_EXPAND:
-    case PRIM_QUERY:
-    case PRIM_QUERY_PARAM_FIELD:
-    case PRIM_QUERY_TYPE_FIELD:
-    case PRIM_OPTIMIZE_ARRAY_BLK_MULT:
-    case PRIM_ACTUALS_LIST:
-    case PRIM_YIELD:
-
-    case PRIM_USED_MODULES_LIST:
-
-    case PRIM_WHEN:
-    case PRIM_CAPTURE_FN_FOR_C:
-    case PRIM_CAPTURE_FN_FOR_CHPL:
-    case PRIM_CREATE_FN_TYPE:
-
-    case PRIM_NUM_FIELDS:
-    case PRIM_IS_POD:
-    case PRIM_FIELD_NUM_TO_NAME:
-    case PRIM_FIELD_NAME_TO_NUM:
-    case PRIM_FIELD_BY_NUM:
-
-    case PRIM_TO_STANDALONE:
-    case PRIM_IS_REF_ITER_TYPE:
-    case PRIM_COERCE:
-    case PRIM_CALL_RESOLVES:
-    case PRIM_METHOD_CALL_RESOLVES:
-    case PRIM_GET_COMPILER_VAR:
-    case PRIM_ZIP:
-    case PRIM_REQUIRE:
-    case NUM_KNOWN_PRIMS:
-    case PRIM_ITERATOR_RECORD_FIELD_VALUE_BY_FORMAL:
-    case PRIM_BLOCK_WHILEDO_LOOP:
-    case PRIM_BLOCK_DOWHILE_LOOP:
-    case PRIM_BLOCK_FOR_LOOP:
-    case PRIM_BLOCK_C_FOR_LOOP:
-    case PRIM_ARRAY_ALLOC:
-    case PRIM_ARRAY_FREE:
-    case PRIM_ARRAY_FREE_ELTS:
-    case PRIM_STRING_COPY:
-    case PRIM_GET_END_COUNT:
-    case PRIM_SET_END_COUNT:
-    case PRIM_TO_LEADER:
-    case PRIM_TO_FOLLOWER:
-    case PRIM_DELETE:
-    case PRIM_CALL_DESTRUCTOR:
-    case PRIM_HEAP_REGISTER_GLOBAL_VAR:
-    case PRIM_HEAP_BROADCAST_GLOBAL_VARS:
-    case PRIM_PRIVATE_BROADCAST:
-    case PRIM_RT_ERROR:
-    case PRIM_RT_WARNING:
-    case PRIM_FTABLE_CALL:
-    case PRIM_VIRTUAL_METHOD_CALL:
-    case PRIM_INT_ERROR:
+    default:
       return false;
   }
-  return false;
+}
+
+//
+// Assumes 'parent' is a PRIM_MOVE or PRIM_ASSIGN
+//
+// Returns false if the LHS of the move/assign indicates that the rhs cannot
+// be a const-ref. For example, if we have a case like this:
+//
+// (move A, (set-reference B))
+//
+// where 'A' and 'B' are references, B cannot be a const-ref if A is not a
+// const-ref.
+//
+// In the case of a dereference, (move A, (deref B)), this function will return
+// true because we're simply reading B.
+//
+static bool canRHSBeConstRef(CallExpr* parent, SymExpr* use) {
+  INT_ASSERT(isMoveOrAssign(parent));
+  SymExpr* LHS = toSymExpr(parent->get(1));
+  CallExpr* rhs = toCallExpr(parent->get(2));
+  INT_ASSERT(rhs);
+  switch (rhs->primitive->tag) {
+    case PRIM_GET_MEMBER:
+    case PRIM_GET_MEMBER_VALUE:
+    case PRIM_GET_SVEC_MEMBER:
+    case PRIM_GET_SVEC_MEMBER_VALUE:
+    case PRIM_GET_REAL:
+    case PRIM_GET_IMAG:
+    case PRIM_ADDR_OF:
+    case PRIM_SET_REFERENCE: {
+      // If LHS is a reference and is not a const-ref, the reference in 'rhs'
+      // should not be considered a const-ref either.
+      //
+      // For the get-member primitives, I intend this to be a safe approach
+      // until we know what const-ref means for fields. Basically, if any field
+      // might be modified I do not consider the base object to be const-ref.
```

Comment:
If that's the case, why disqualify PRIM_GET_MEMBER_VALUE / PRIM_GET_SVEC_MEMBER_VALUE ? Since these do not return a reference. (Or maybe they do in the case of a reference field? Oh, I see, and then you check if LHS is a reference below? I think it'd be worth commenting on this since it's not obvious at a glance.